File: server/services/simpleAiDetector.ts
export interface AIDetectionResult {
  probability: number;
  label: 'AI Generated' | 'Human Written';
  confidence: number;
  miraAnalysis: string;
}

export class SimpleAIDetectorService {
  async detectAIText(inputText: string): Promise<AIDetectionResult> {
    console.log('Running DeefakeTextDetection Python script (user provided algorithm)...');
    
    // ALWAYS use the user's Python script - NO FALLBACKS
    const pythonResult = await this.runPythonDetector(inputText);
    
    if (pythonResult.error) {
      throw new Error(`DeefakeTextDetection script failed: ${pythonResult.error}`);
    }

    const miraAnalysis = this.generateSimpleMiraAnalysis(inputText, pythonResult.probability, pythonResult.label);
    
    return {
      probability: pythonResult.probability,
      label: pythonResult.label,
      confidence: this.calculateConfidence(pythonResult.probability),
      miraAnalysis
    };
  }

  private async runPythonDetector(text: string): Promise<{ probability: number; label: 'AI Generated' | 'Human Written'; error?: string }> {
    return new Promise((resolve) => {
      const { spawn } = require('child_process');
      const python = spawn('python3', ['server/services/ai-detector.py', text]);
      
      let output = '';
      let errorOutput = '';

      python.stdout.on('data', (data: any) => {
        output += data.toString();
      });

      python.stderr.on('data', (data: any) => {
        errorOutput += data.toString();
      });

      python.on('close', (code: number) => {
        if (code !== 0) {
          console.error('DeefakeTextDetection script error:', errorOutput);
          resolve({ probability: 0.5, label: 'AI Generated', error: errorOutput });
          return;
        }

        try {
          const result = JSON.parse(output.trim());
          console.log(`DeefakeTextDetection result: ${(result.probability * 100).toFixed(1)}% AI probability (${result.label})`);
          resolve({
            probability: result.probability,
            label: result.label as 'AI Generated' | 'Human Written'
          });
        } catch (parseError) {
          console.error('JSON parse error:', parseError);
          resolve({ probability: 0.5, label: 'AI Generated', error: 'Failed to parse result' });
        }
      });
    });
  }

  private analyzeTextCharacteristics(text: string): {probability: number, label: 'AI Generated' | 'Human Written'} {
    const textLength = text.length;
    const wordCount = text.split(/\s+/).filter(word => word.length > 0).length;
    const avgWordLength = text.replace(/[^a-zA-Z]/g, '').length / Math.max(wordCount, 1);
    const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
    const avgSentenceLength = wordCount / Math.max(sentences.length, 1);
    
    let aiProbability = 0.5; // Start neutral
    
    // Structural analysis
    if (avgSentenceLength > 25) aiProbability += 0.15; // Very long sentences
    if (avgWordLength > 5.5) aiProbability += 0.1; // Complex vocabulary
    if (sentences.length > 2 && sentences.every(s => s.trim().length > 10)) aiProbability += 0.1; // Consistent length
    
    // AI-typical patterns
    const aiPatterns = [
      /\b(furthermore|moreover|additionally|consequently|therefore|however)\b/gi,
      /\b(it's worth noting|it's important to|as an AI|I should mention)\b/gi,
      /\b(in conclusion|to summarize|in summary|overall)\b/gi,
      /\b(comprehensive|extensive|various|numerous|significant)\b/gi
    ];
    
    aiPatterns.forEach(pattern => {
      const matches = text.match(pattern);
      if (matches) {
        aiProbability += matches.length * 0.08;
      }
    });
    
    // Human-typical patterns
    const humanPatterns = [
      /\b(uh|um|like|you know|I mean|actually|basically)\b/gi,
      /\b(lol|haha|omg|btw|tbh|imo|fyi)\b/gi,
      /[.]{3,}|\?\?\?|!!!+|wow|cool|awesome/gi,
      /\b(I think|I feel|I guess|maybe|probably)\b/gi
    ];
    
    humanPatterns.forEach(pattern => {
      const matches = text.match(pattern);
      if (matches) {
        aiProbability -= matches.length * 0.12;
      }
    });
    
    // Length-based adjustments
    if (wordCount < 20) aiProbability -= 0.1; // Short texts often human
    if (wordCount > 200) aiProbability += 0.05; // Long formal texts often AI
    
    // Punctuation analysis
    const exclamations = (text.match(/!/g) || []).length;
    const questions = (text.match(/\?/g) || []).length;
    if (exclamations + questions > wordCount * 0.05) aiProbability -= 0.1; // Emotional punctuation
    
    // Vocabulary complexity
    const uniqueWords = new Set(text.toLowerCase().match(/\b\w+\b/g) || []);
    const vocabularyRichness = uniqueWords.size / Math.max(wordCount, 1);
    if (vocabularyRichness > 0.7) aiProbability += 0.1; // Very diverse vocabulary
    
    // Personal pronouns (humans use more first person)
    const firstPersonCount = (text.match(/\b(I|me|my|mine|myself)\b/gi) || []).length;
    const firstPersonRatio = firstPersonCount / Math.max(wordCount, 1);
    if (firstPersonRatio > 0.05) aiProbability -= 0.1;
    
    // Add controlled randomness for realism
    const randomFactor = (Math.random() - 0.5) * 0.1;
    aiProbability += randomFactor;
    
    // Clamp probability between 0.05 and 0.95 for realism
    aiProbability = Math.max(0.05, Math.min(0.95, aiProbability));
    
    const label: 'AI Generated' | 'Human Written' = aiProbability > 0.5 ? 'AI Generated' : 'Human Written';
    
    console.log(`Text analysis complete: ${(aiProbability * 100).toFixed(1)}% AI probability (${label})`);
    
    return {
      probability: aiProbability,
      label
    };
  }

  private calculateConfidence(probability: number): number {
    // Calculate confidence based on how far the probability is from 0.5 (uncertain)
    const distance = Math.abs(probability - 0.5);
    return Math.min(distance * 2, 1); // Scale to 0-1 range
  }

  private generateSimpleMiraAnalysis(text: string, probability: number, label: string): string {
    const textPreview = text.length > 50 ? text.substring(0, 50) + "..." : text;
    const probPercent = (probability * 100).toFixed(1);
    
    const responses = [
      `Yo! I've scanned "${textPreview}" and my cybernetic neural nets are saying ${probPercent}% chance this is ${label.toLowerCase()}. The writing patterns are pretty telling!`,
      `Analyzed your text snippet there. My detection algorithms show ${probPercent}% probability of being ${label.toLowerCase()}. The structure and word choices give it away.`,
      `Just ran my AI detection protocols on that text. Getting ${probPercent}% confidence it's ${label.toLowerCase()}. The linguistic fingerprints are quite clear.`,
      `Processed that text sample through my detection matrix. Results: ${probPercent}% likelihood of ${label.toLowerCase()}. The patterns are distinctive!`,
      `My analysis engine churned through your text and found ${probPercent}% probability it's ${label.toLowerCase()}. The stylistic markers tell the story.`
    ];
    
    return responses[Math.floor(Math.random() * responses.length)];
  }
}

export const simpleAiDetectorService = new SimpleAIDetectorService();

File: server/services/simple-ai-detector.py
#!/usr/bin/env python3
import sys
import json
import re
import math

def analyze_text_patterns(text):
    """Advanced linguistic pattern analysis for AI detection"""
    
    # Calculate various linguistic features
    features = {}
    
    # Basic text statistics
    words = text.split()
    sentences = re.split(r'[.!?]+', text)
    sentences = [s.strip() for s in sentences if s.strip()]
    
    features['word_count'] = len(words)
    features['sentence_count'] = len(sentences)
    features['avg_word_length'] = sum(len(word) for word in words) / max(len(words), 1)
    features['avg_sentence_length'] = len(words) / max(len(sentences), 1)
    
    # AI-generated text patterns
    ai_indicators = 0
    total_checks = 0
    
    # 1. Formal language patterns
    formal_phrases = [
        'furthermore', 'additionally', 'moreover', 'however', 'therefore',
        'comprehensive', 'extensive', 'significant', 'substantial', 'considerable',
        'analysis', 'findings', 'implications', 'methodological', 'demonstrates',
        'it is important to note', 'research suggests', 'studies indicate'
    ]
    
    formal_count = sum(1 for phrase in formal_phrases if phrase.lower() in text.lower())
    formal_density = formal_count / max(len(words), 1) * 100
    
    if formal_density > 5:  # High formal phrase density
        ai_indicators += 2
    elif formal_density > 2:
        ai_indicators += 1
    total_checks += 2
    
    # 2. Sentence structure consistency (AI tends to be very consistent)
    sentence_lengths = [len(s.split()) for s in sentences if s.strip()]
    if len(sentence_lengths) > 1:
        length_variance = sum((x - features['avg_sentence_length'])**2 for x in sentence_lengths) / len(sentence_lengths)
        if length_variance < 10:  # Very consistent sentence lengths
            ai_indicators += 1
        total_checks += 1
    
    # 3. Lack of contractions (AI often avoids contractions)
    contractions = ["don't", "won't", "can't", "shouldn't", "wouldn't", "couldn't", "isn't", "aren't", "wasn't", "weren't", "i'm", "you're", "we're", "they're", "i've", "you've", "we've", "they've", "i'll", "you'll", "we'll", "they'll"]
    contraction_count = sum(1 for contraction in contractions if contraction.lower() in text.lower())
    contraction_density = contraction_count / max(len(words), 1) * 100
    
    if contraction_density < 0.5:  # Very few contractions
        ai_indicators += 1
    total_checks += 1
    
    # 4. Redundant phrasing patterns
    redundant_patterns = [
        r'it is (important|crucial|essential|vital) to (note|understand|recognize|consider)',
        r'(furthermore|additionally|moreover), it (should be|is) (noted|mentioned|emphasized)',
        r'(extensive|comprehensive|significant|substantial) (research|analysis|study|investigation)'
    ]
    
    redundant_matches = 0
    for pattern in redundant_patterns:
        if re.search(pattern, text, re.IGNORECASE):
            redundant_matches += 1
    
    if redundant_matches >= 2:
        ai_indicators += 2
    elif redundant_matches >= 1:
        ai_indicators += 1
    total_checks += 2
    
    # 5. Perfect grammar with complex vocabulary (unusual for casual text)
    complex_words = sum(1 for word in words if len(word) > 8)
    complex_word_ratio = complex_words / max(len(words), 1)
    
    # Check for grammar errors (simple heuristics)
    grammar_errors = 0
    if not re.search(r'[.!?]$', text.strip()):  # No ending punctuation
        grammar_errors += 1
    if re.search(r'\s+[.!?]', text):  # Space before punctuation
        grammar_errors += 1
    if re.search(r'[a-z][A-Z]', text):  # Inconsistent capitalization
        grammar_errors += 1
    
    if complex_word_ratio > 0.15 and grammar_errors == 0:  # Complex words + perfect grammar
        ai_indicators += 1
    total_checks += 1
    
    # 6. Casual language detection (more human-like)
    casual_indicators = [
        'lol', 'omg', 'wow', 'cool', 'awesome', 'pretty', 'really', 'super',
        'like', 'kinda', 'sorta', 'gonna', 'wanna', 'yeah', 'nah', 'btw'
    ]
    
    casual_count = sum(1 for indicator in casual_indicators if indicator.lower() in text.lower())
    casual_density = casual_count / max(len(words), 1) * 100
    
    if casual_density > 2:  # High casual language
        ai_indicators -= 1  # Reduces AI probability
    total_checks += 1
    
    # Calculate final probability
    if total_checks > 0:
        ai_probability = min(0.95, max(0.05, ai_indicators / total_checks))
    else:
        ai_probability = 0.5  # Default if no analysis possible
    
    # Adjust based on text length (very short texts are harder to classify)
    if len(words) < 20:
        ai_probability = ai_probability * 0.8 + 0.1  # Less confident for short texts
    
    # Determine label and confidence
    if ai_probability > 0.7:
        label = "AI Generated"
        confidence = ai_probability
    elif ai_probability < 0.3:
        label = "Human Written"
        confidence = 1 - ai_probability
    else:
        label = "AI Generated" if ai_probability >= 0.5 else "Human Written"
        confidence = max(ai_probability, 1 - ai_probability) * 0.7  # Lower confidence for borderline cases
    
    return {
        "probability": ai_probability,
        "label": label,
        "confidence": confidence,
        "features": {
            "formal_density": formal_density,
            "avg_sentence_length": features['avg_sentence_length'],
            "contraction_density": contraction_density,
            "complex_word_ratio": complex_word_ratio,
            "casual_density": casual_density
        }
    }

def main():
    try:
        # Read input text from stdin
        input_text = sys.stdin.read().strip()
        
        if not input_text:
            print(json.dumps({"error": "No input text provided"}))
            sys.exit(1)
        
        # Analyze the text
        result = analyze_text_patterns(input_text)
        
        # Output JSON result
        print(json.dumps(result))
        
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

if __name__ == "__main__":
    main()